File: 1114-print-in-order/1114-print-in-order.cpp
#include <semaphore.h>
class Foo {
public:
    sem_t p1;
    sem_t p2;
    Foo() {
        sem_init(&p1,0,0);
        sem_init(&p2,0,0);
        
    }

    void first(function<void()> printFirst) {
        
        // printFirst() outputs "first". Do not change or remove this line.
        printFirst();
        sem_post(&p1);
    }

    void second(function<void()> printSecond) {
        sem_wait(&p1);
        // printSecond() outputs "second". Do not change or remove this line.
        printSecond();
        sem_post(&p2);
    }

    void third(function<void()> printThird) {
        sem_wait(&p2);
        // printThird() outputs "third". Do not change or remove this line.
        printThird();
    }
};

File: 1114-print-in-order/1114-print-in-order.py
import threading
class Foo:
    def __init__(self):
        self.p1=threading.Lock()
        self.p2=threading.Lock()
        self.p1.acquire()
        self.p2.acquire()


    def first(self, printFirst: 'Callable[[], None]') -> None:
        # printFirst() outputs "first". Do not change or remove this line.
        printFirst()
        self.p1.release()


    def second(self, printSecond: 'Callable[[], None]') -> None:
        self.p1.acquire()
        # printSecond() outputs "second". Do not change or remove this line.
        printSecond()
        self.p2.release()

    def third(self, printThird: 'Callable[[], None]') -> None:
        self.p2.acquire()
        # printThird() outputs "third". Do not change or remove this line.
        printThird()